Added tests for DatabaseFileLookup

arfazhxss opened the pull request:
Closes https://github.com/koppor/jabref/issues/678

I have added new test cases directoryPathTests to the DatabaseFileLookupTest class. These test cases verifies the functionality of the DatabaseFileLookup class in handling file paths and directories. It creates a temporary directory and file, sets up a BibDatabaseContext with two entries (one with a valid file path and one with an empty file path), and configures the FilePreferences to use the temporary directory as the default file directory. Then, it creates an instance of DatabaseFileLookup and asserts that the file in the temporary directory is found, while a non-existent file is not found. The test case checks if the FilePreferences are correctly set up with the temporary directory as the main file directory, and if the getPathOfDatabase() method of DatabaseFileLookup returns an empty path.

### Mandatory checks

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor commented:
Please first fix tests to enable easier reviewing. You see the failing tests at the red crosses of the page of this PR. @arfazhxss

koppor commented:
I think, AuthirListTest should not be part of the PR

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -44,4 +59,50 @@ void prerequisitesFulfilled() {
+
+        // Set the temporary directory as the default file directory
+        // in the preferences and creating DatabaseFileLookup instance
+        FilePreferences filePreferences = new FilePreferences("", txtFileDir.toAbsolutePath().toString(), false, "", "", false, false, null, Collections.emptySet(), false, null);
```
Please mock the FilePreferences. See other test classes

arfazhxss replied:
@koppor I have mocked the FilePreferences, along with other functions except the DatabaseFileLookup class, as we are testing that class. Let me know if I need to change anything in my implementation.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -44,4 +59,50 @@ void prerequisitesFulfilled() {
+        // Create a BibDatabaseContext with a BibDatabase containing two entries
+        BibDatabase bibDatabase = new BibDatabase();
+        BibEntry entry1 = new BibEntry();
+        entry1.setField(StandardField.FILE, txtFileDir.toAbsolutePath().toString());
```
Use .withX...  - see other test classes

arfazhxss replied:
@koppor I have rewrote the entire test file into 3 parts: 

[x] SetUp(): Preparing the Database environments, entries, mockups and required dependencies for the DatabaseFileLookup class, 
[x] FileShouldBeFound() with the tests checking if the mocked database entry for the file "x.txt" with the temporary directory matches the intended DatabaseFileLookup directory. 
[x] fileShouldNotBeFound with the tests checking if the database has the location for the newly resolved entry of the file "y.txt"
[ ] PrerequisitesFulfilled(): this is a test file that checks the prerequisites for the tests to be carried out.

Let me know if I can improve/add more tests to this.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -44,4 +59,50 @@ void prerequisitesFulfilled() {
+        entry1.setField(StandardField.FILE, txtFileDir.toAbsolutePath().toString());
+        BibEntry entry2 = new BibEntry();
+        entry2.setField(StandardField.FILE, "");
+        bibDatabase.insertEntry(entry1);
```
See other test classes. Pass a list of entries to constructor

arfazhxss replied:
As mentioned in one my previous comment, I have used SetUp() to prepare the constructors for all the tests. If there's space for improving the tests, through further introducing constructors or dividing up the SetUp(), or making the code more readable in any way, let me know!

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -44,4 +59,50 @@ void prerequisitesFulfilled() {
+        FilePreferences filePreferences = new FilePreferences("", txtFileDir.toAbsolutePath().toString(), false, "", "", false, false, null, Collections.emptySet(), false, null);
+        DatabaseFileLookup fileLookup = new DatabaseFileLookup(databaseContext, filePreferences);
+
+        assertTrue(fileLookup.lookupDatabase(txtFileDir)); // x.txt should be found
```
One test case, one assetion.

Separate tests

arfazhxss replied:
I have made 3 assertions for the FileShouldBeFound test. I can remove one of the assertions that checks if the database is in the intended directory and put it into a separate test case. Other than that, I have separated all the assertions based on specific test cases.

koppor reviewed (changes requested):
This test is a good start. It misses testing of different configurated file directies. Please re-iterate in the documentation.

In numbers: 20% of the issue solved.

arfazhxss commented:
> @arfazhxss in future, do NOT use the main branch for cide changes. Create a separate branch for each issue solved!

Thank you for the advice:) I am working on fixing the errors from the tests.

arfazhxss commented:
@Siedlerchr I have one failing checks for the checkStyle for now:

- [ ] DatabaseFileLookupTest.java:10:1: Wrong order for 'java.nio.file.Files' import. [ImportOrder]
- [ ] DatabaseFileLookupTest.java:18:1: 'javafx.collections.FXCollections' should be separated from previous imports. [ImportOrder]
- [ ] DatabaseFileLookupTest.java:20:1: 'org.jabref.logic.importer.fileformat.BibtexImporter' should be separated from previous imports. [ImportOrder]
- [ ] DatabaseFileLookupTest.java:28:1: 'org.junit.jupiter.api.BeforeEach' should be separated from previous imports. [ImportOrder]

I think I might need a bit help on this one.

arfazhxss commented:
![IntelliJ IDE Coverage](https://github.com/JabRef/jabref/assets/84450659/b1837b9c-70d2-47ae-b4c4-4769f38b305d)

For now, the tests are covering 100% of classes, lines and methods. For the branch, the tests haven't been added for the .filter() method in the parseFileField() function.

Siedlerchr commented:
In Intellij menu -> Code Reformat code  and optimize imports

Siedlerchr commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,134 @@
+    BibEntry entry4 = new BibEntry().withField(StandardField.FILE, "");
+
+    List<BibEntry> entries = new ArrayList<>(
+      Arrays.asList(entry1, entry2, entry3, entry4)
```
Just List.of(entry1, entry2...)
But beware that List.of creates a new unmodifiable list.
If you need a modifable just   List<BIbEntry> = Arrays.asList(entry1, entry2, entry3, entry4)

koppor commented on CHANGELOG.md:
```diff
@@ -59,6 +59,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - The XMP-Exporter no longer writes the content of the `file`-field. [#11083](https://github.com/JabRef/jabref/pull/11083)
 - We added notes, checks and warnings for the case of selection of non-empty directories while starting a new Systematic Literature Review. [#600](https://github.com/koppor/jabref/issues/600)
 - Text in the import dialog (web search results) will now be wrapped to prevent horizontal scrolling. [#10931](https://github.com/JabRef/jabref/issues/10931)
+- We added test cases for DatabaseFileLookup to cover file presence and absence scenarios. [koppor#678](https://github.com/koppor/jabref/issues/678)
```
CHANGELOG is for user-facing issues only. Thus, remove the entry.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -22,16 +38,52 @@ class DatabaseFileLookupTest {
+    private FilePreferences filePreferences;
+    private DatabaseFileLookup fileLookup;
 
+    /**
```
The comment can be derived from the method name and signature. Thus, remove to keep the LOC count low.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -22,16 +38,52 @@ class DatabaseFileLookupTest {
         database = result.getDatabase();
         entries = database.getEntries();
 
+        tempDir = Files.createTempDirectory("testDir");
```
Use @TempDir from JUnit

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -22,16 +38,52 @@ class DatabaseFileLookupTest {
+        BibEntry entry4 = new BibEntry().withField(StandardField.FILE, "");
+
+//        List<BibEntry> entries = new ArrayList<>(Arrays.asList(entry1, entry2, entry3, entry4));
+        List<BibEntry> entries = Arrays.asList(entry1, entry2, entry3, entry4);
```
Use List.of

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -22,16 +38,52 @@ class DatabaseFileLookupTest {
+        List<BibEntry> entries = Arrays.asList(entry1, entry2, entry3, entry4);
+        ObservableList<BibEntry> observableEntryList = FXCollections
+                .synchronizedObservableList(FXCollections.observableArrayList(entries));
+        BibDatabase databaseMock = mock(BibDatabase.class);
```
You can use the real BibDatabase class, because the lookup tests test also that class

koppor reviewed (changes requested):
Thank you for working on this

It does not fulfill the aims of the issue. JabRef has four places to store PDFs: Next to the bib file, in the general file directory, database file directory, the user-specific file directory. I don't see tests for all these cases.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -44,4 +96,23 @@ void prerequisitesFulfilled() {
+    @Test
+    void fileShouldBeFound() {
+        assertTrue(fileLookup.lookupDatabase(txtFileDir));
+        assertEquals(filePreferences.getMainFileDirectory().orElse(Path.of("")).toString(),
```
Compare Optional.of(Path) directly

arfazhxss commented:
I will keep working on this pull request throughout next couple of weeks

arfazhxss commented:
In my latest changes, I fixed the following requested changes:

- [x] Removed redundant comments
- [x] Added @tempdir from JUnit
- [x] Modified List.of from Array Implementation
- [x] Using the actual BibDatabase, BibDatabaseContext and Metadata Implementation
- [x] One assertion per test
- [x] CheckStyle corrections, optimized imports and code reformating
- [x] Added test cases for general file directory, database file directory, the user-specific file directory and bib database file directory

Please let me know how I can improve the test cases if needed.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,146 @@
 
-        entry1 = database.getEntryByCitationKey("entry1").get();
-        entry2 = database.getEntryByCitationKey("entry2").get();
+        tempDir = Files.createTempDirectory("testDir");
```
Please use JUnit `@TempDir` and not Files.createTEmpDirectory

arfazhxss replied:
@koppor I replaced the manual creation of a temporary directory with JUnit's @TempDir annotation for automatic handling.

koppor reviewed (changes requested):
I am so sorry that my "hints" section at https://github.com/koppor/jabref/issues/678 made the impression that the task list was complete.

The intention of the task was that students think of test cases.

Here some hints:

- JabRef has 3 different file directories
  - global one
  - library-specific
  - user-specific
 
Thus, file can be placed in


- global one
- library-specific
- user-specific
- directory of other user

File can exist in multiple directories.

The work should cover all cases derived from that.

E.g., 

- file in user-specific directory and global directory. Found in user-specific dir.
- file in user-specific directory and library-specific directory. Found in user-specific dir.
- file in user-specific directory and library-specific directory and global directory. Found in user-specific dir.
- file in user-specific directory of other user and library-specific directory  and global directory. Found in library-specific dir.
- file in user-specific directory of other user and global directory. Found in global dir.
- file in library-specific directory and global directory. Found in library-specific dir.
- File in global-directory only. Found there.
- File in user-specific directory. Found there.
- File in library-specific directory. Found there.

---

This what I meant that 20% of the work have been done.

arfazhxss commented:
> * JabRef has 3 different file directories
>   
>   * global one
>   * library-specific
>   * user-specific
> 
> Thus, file can be placed in
> 
> * global one
> * library-specific
> * user-specific
> * directory of other user
> 
> File can exist in multiple directories.
> 
> The work should cover all cases derived from that.

@koppor thanks for clarifying! I want to get some more clarification on my test cases (and how I can improve them). My current test cases makes an x.txt  and passes it through the metadata and databaseContext, and then uses DatabaseFileLookup to test whether the file is correctly detected in the database. It also checks if that file exists in the general file directory, database file directory, the user-specific file directory and bib/library file directory. Does it mean that my current test cases covers the following marked-cases only for now?

- [ ] file in user-specific directory and global directory. Found in user-specific dir.
- [ ] file in user-specific directory and library-specific directory. Found in user-specific dir.
- [ ] file in user-specific directory and library-specific directory and global directory. Found in user-specific dir.
- [ ] file in user-specific directory of other user and library-specific directory and global directory. Found in library-specific dir.
- [ ] file in user-specific directory of other user and global directory. Found in global dir.
- [ ] file in library-specific directory and global directory. Found in library-specific dir.
- [x] File in global-directory only. Found there.
- [x] File in user-specific directory. Found there.
- [x] File in library-specific directory. Found there.

This is just to clarify if I am on the right track.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,145 @@
     void prerequisitesFulfilled() {
-        assertEquals(2, database.getEntryCount());
+        assertEquals(4, database.getEntryCount());
         assertEquals(2, entries.size());
         assertNotNull(entry1);
         assertNotNull(entry2);
```
A test case should focus on **one** thing. Not testing that constructors of data classes work. Remove all three assertions.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,145 @@
+     */
+    @Test
+    void fileShouldBeFound() {
+        assertTrue(fileLookup.lookupDatabase(txtFileDir));
```
This assertion is OK.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,145 @@
+    @Test
+    void fileShouldBeFound() {
+        assertTrue(fileLookup.lookupDatabase(txtFileDir));
+        assertNotNull(fileLookup.getPathOfDatabase());
```
Remove this. Tests other aspects.

koppor commented on src/test/java/org/jabref/logic/importer/DatabaseFileLookupTest.java:
```diff
@@ -1,47 +1,145 @@
+        assertFalse(fileLookup.lookupDatabase(tempDir.resolve("y.txt")));
+    }
+
+    /**
+     * x.txt should be found in the user-specific directory
+     */
+    @Test
+    void userSpecificDirectory() {
+        assertEquals(metaData.getUserFileDirectory(filePreferences.getUserAndHost()), Optional.of(txtFileDirPath));
+    }
```
description does not match assertion (does it)?

koppor reviewed (changes requested):
Your whole tests need to be reorganized. Sorry for this.

One test for case.

Example test names and content:

- `librarySpecificDirectoryOnly`. Places the file in library-specific directory. Checks that this file is found.
- `userSpecificDirectoryAndLibrarySpecificDirectory`. Pleases the file in both user-specific and library-specific directory. Asserts that the file is found in the user-specific directory.

Note that the file is stored **as filename without path**.

Thus `txtFileDirPath = txtFileDir.toAbsolutePath()` is absolutely wrong!!

The part of the BibEntry is `file = {:test.pdf}`. And JabRef searches for that file.

(If you use absolute paths, it is boring, because JabRef can also ignore the directories, because the path is absolute. It is about **resolving**)

koppor commented:
@arfazhxss I refined the issue to be more clear. Look at "User-facing thing" at https://github.com/koppor/jabref/issues/678. Hope, that helps to understand the issue.

koppor commented:
> My current test cases makes an x.txt

As far as I  understand, you put an **absolute** path. Then, nothing of JabRef's lookup functionality is triggered. Story `x.txt` as is in `StandardFields.File` and see what happens.

koppor commented:
@arfazhxss I did not see any activity the last weeks. We need to keep our number of opened pull requests low to enable maintaining the project. If you continue on this, feel free to open a new pull request.
